security: Fix command injection vulnerabilities

sm4rtdev opened the pull request:
## Changes

This PR fixes multiple critical command injection vulnerabilities across the codebase by implementing proper input sanitization and escaping.

### Key Changes:

1. **Command Injection Fixes**
   - Applied `shlex.quote()` to all user-controlled inputs in shell commands
   - Fixed vulnerabilities in SSH operations (`interactive_shell_service.py`)
   - Fixed vulnerabilities in Docker operations (`docker_service.py`)
   - Fixed vulnerabilities in VerifyX validation (`verifyx_validation_service.py`)
   - Fixed vulnerabilities in executor connectivity (`executor_connectivity_service.py`)
   - Fixed vulnerabilities in rented machine checks (`rented_machine.py`)

2. **Subprocess Security**
   - Removed `shell=True` from all subprocess calls to prevent shell injection
   - Converted command strings to command lists for safer execution
   - Applied `shlex.split()` for safe command parsing
   - Fixed in: `hash_service.py`, `score.py`, `machine_scrape.py`, `backup_storage.py`, `restore_storage.py`

3. **Input Sanitization**
   - Added `_sanitize_startup_commands()` method to filter dangerous shell metacharacters
   - Enhanced `CustomOptions.sanitize()` to include startup_commands validation

### Security Impact:
- **Before:** Attackers could potentially execute arbitrary commands by injecting malicious input
- **After:** All user inputs are properly escaped and validated, preventing command injection attacks

### Testing:
- All changes maintain backward compatibility
- Functionality remains unchanged
- No breaking changes introduced

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I wrote tests.
- [x] Need to take care of performance?

### Notes:
- **Tests:** Security fixes maintain existing functionality. Integration tests should verify that command execution still works correctly with escaped inputs.
- **Performance:** Minimal impact. `shlex.quote()` and `shlex.split()` are lightweight operations. No significant performance degradation expected.

pyon12 committed:
fix(validator): DAH-1590 - gpu flag in docker service

arhangel66 committed:
validator: use saved port mapping validation and test updates

arhangel66 committed:
validator: add logs to see what is happening.

arhangel66 committed:
validator: added migration

arhangel66 committed:
validator: fix migration

arhangel66 committed:
validator: fix payload.pod_id

arhangel66 committed:
validator: move pod_id to be required param

LegendWorrier committed:
chore: add doc links in subnet readme

arhangel66 committed:
validator: added lock on choose port logic

LegendWorrier committed:
chore: add doc links in all readme

pyon12 committed:
Merge pull request #802 from Datura-ai/chore/add-doc-link-subnet-readme

chore: add doc links in subnet readme

LegendWorrier committed:
Merge branch 'main' into feat/DAH-1572-pod-utilization-pod-logs

LegendWorrier committed:
chore: remove pod_id in pod utilitzation endpoint

LegendWorrier committed:
chore: update pod utilization enpoint again

LegendWorrier committed:
fix: remove pod_id in middleware of miner

arhangel66 committed:
validator: improve timeout to not allow miner_timeout

pyon12 committed:
fix: timeoutError

pyon12 committed:
Merge pull request #803 from Datura-ai/fix/miner-timeout-issue

fix(validator): adjust timeouts and add execution metrics

pyon12 committed:
Update version to 3.13.6

pyon12 committed:
fix(executor,validator): fixed executor dockerfile issue and update machine_scrape for speed commands

pyon12 committed:
fix: potential issues

LegendWorrier committed:
Merge branch 'main' into feat/DAH-1572-pod-utilization-pod-logs

pyon12 committed:
fix: merge conflict from port mapping branch

pyon12 committed:
feat(validator): DAH-1589 - optional port_connectivity check

pyon12 committed:
fix(validator): DAH-1589 - cleaning existing containers when renting a pod

pyon12 committed:
fix(validator): DAH-1589 - exclude rented ports from available_port_count

arhangel66 committed:
added pool_pre_ping and remove wrong session.commit() from upsert_ports.

pyon12 committed:
Merge pull request #805 from Datura-ai/fix/executor-dockerfile

fix(executor,validator): fixed executor dockerfile issue

pyon12 committed:
chore: bump versions

pyon12 committed:
fix(validator): backup/restore issue

pyon12 committed:
Merge pull request #809 from Datura-ai/fix/executor-dockerfile

fix(validator): backup/restore issue

pyon12 committed:
chore(validator): bump version

Commit:
Merge branch 'main' into feat/DAH-1572-pod-utilization-pod-logs

pyon12 committed:
Merge pull request #780 from Datura-ai/feat/DAH-1572-pod-utilization-pod-logs

feat: DAH-1572 - pod utilization and pod logs

pyon12 committed:
chore: bump versions

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

LegendWorrier committed:
feat: update /containers/{container_name} endpoint by adding get_container_metrics function

slickroot committed:
fix(subnet): upgrading to a more maintained watchtower repository

slickroot committed:
feat(subnet): Miners can see that we couldn't reach their miner proxy or any other problems that happen while requesting the job

arhangel66 committed:
remove 2 useless commits + extend session config.

pyon12 committed:
Merge pull request #814 from Datura-ai/fix/DAH-1597-watchtower-deprecated-docker-29

fix(subnet): upgrading to a more maintained watchtower repository

pyon12 committed:
Merge pull request #801 from Datura-ai/feature/save-port-map

validator: use saved port mapping validation and test updates

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

slickroot committed:
fix(subnet): Publish even when no specs

pyon12 committed:
Merge pull request #813 from Datura-ai/feat/DAH-1590-pod-utilization

feat: update /containers/{container_name} endpoint

pyon12 committed:
chore(executor): bump version

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

arhangel66 committed:
(validator) fix filtering rented ports.

arhangel66 committed:
(validator) fix filtering rented ports.

fortunelucky777 committed:
Merge pull request #818 from Datura-ai/fix/validator-filter-busy-ports

(validator) fix filtering rented ports.

fortunelucky777 committed:
chore(validator): bump version to 3.13.11

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

pyon12 committed:
hotfix: error log for matrix verification

pyon12 committed:
chore(validator): bump version

slickroot committed:
feat(subnet): skip rental check online, do not rely on systemd or init being on the container, run sshd directly

pyon12 committed:
Merge pull request #822 from Datura-ai/feat/skip-rental-verification-locally

feat(subnet): skip rental check online, do not rely on systemd or ini…

pyon12 committed:
chore(validator): bump version

fortunelucky777 committed:
feat(miner): add debug settings to skip validator registration and sync flow

fortunelucky777 committed:
refactor(validator, miner): update debug settings structure and environment variables for miners and validators

fortunelucky777 committed:
feat(validator): add SKIP_STORAGE_CHECK option to debug settings and conditionally include storage validation

fortunelucky777 committed:
feat(miner): add DEBUG_ENABLED option to environment template and update config for debug mode

slickroot committed:
fix(validator): always install openssh-server instead of broken check

pyon12 committed:
revert(validator): build pipeline

pyon12 committed:
Merge pull request #825 from Datura-ai/fix/openssh-server-false-positive

fix(validator): always install openssh-server instead of broken check

pyon12 committed:
fix(validator): cli update

pyon12 committed:
chore(validator): bump version

pyon12 committed:
fix(validator): conflict and merge master

arhangel66 committed:
feat(validator): DAH-1631 - fail on SSHPubKeyRemoveRequest command will not raise whole miner_task exception.

pyon12 committed:
Merge pull request #827 from Datura-ai/feature/DAH-1631-remove-ssh-key-not-critical

feat(validator): DAH-1631 - make SSHPubKeyRemoveRequest non-blocking for miner task

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

LegendWorrier committed:
feat: update ws conection to restful api between validator and miner

pyon12 committed:
feat(miner): DAH-1632 - remove price_per_hour in sync flow with portal

pyon12 committed:
feat(miner): DAH-1632 - remove price_per_hour in cli

fortunelucky777 committed:
refactor(validator, miner): dry principle & fix security issues

pyon12 committed:
feat(miner): add proper logs for synthetic job, etc

pyon12 committed:
Merge pull request #831 from Datura-ai/hotfix/add-proper-logs-on-miner

feat(miner): add proper logs for synthetic job, etc

pyon12 committed:
chore(miner): bump version

pyon12 committed:
Merge branch 'main' into feat/DAH-1589-gpu-splitting

pyon12 committed:
Merge branch 'main' into feat/DAH-1632-remove-price-per-hour

slickroot committed:
fix(validator): Special UUID for failed miners

slickroot committed:
feat(executor): new authenticated route to stream logs of a docker container

LegendWorrier committed:
chore: cover the comments

pyon12 committed:
Merge pull request #833 from Datura-ai/feat/DAH-1630-stream-logs-from-rented-container

feat(executor): new authenticated route to stream logs of a docker container

pyon12 committed:
Merge pull request #815 from Datura-ai/fix/make-job-result-for-all-miners

feat(subnet): Miners can see that we couldn't reach them

pyon12 committed:
Merge pull request #824 from Datura-ai/refactor/local-dev-flags

refactor(validator, miner): Local development setup

fortunelucky777 committed:
enhance(logging): include exception details and response counts in SSH key operations

fortunelucky777 committed:
Merge branch 'feat/update-ws-connection-to-rest-api-communication' of github.com:Datura-ai/lium-io into feat/update-ws-connection-to-rest-api-communication

fortunelucky777 committed:
Merge branch 'main' into feat/update-ws-connection-to-rest-api-communication

fortunelucky777 committed:
chore: updated gitignore

fortunelucky777 committed:
fix(validator): parsing miner http response & returning failed job result in requesting job

fortunelucky777 committed:
chore(validator): remove unncessary los

LegendWorrier committed:
fix(validator, miner): ensure ssh key removal in handle container error, get rid of removing ssh key on try/catch

fortunelucky777 committed:
"refactor(validator): handling restful api error"

pyon12 committed:
Merge pull request #829 from Datura-ai/feat/update-ws-connection-to-rest-api-communication

feat: update ws conection to restful api between validator and miner

pyon12 committed:
chore: bump versions

pyon12 committed:
chore(executor): bump version

pyon12 committed:
Merge branch 'main' into feat/DAH-1632-remove-price-per-hour

pyon12 committed:
feat: remove price_per_hour from the synthetic job flow

pyon12 committed:
chore: remove unsed DEBUG flag from config

pyon12 committed:
feat(validator): improve job logs when failed

pyon12 committed:
Merge pull request #837 from Datura-ai/fix/validator-failed-job-logs

Fix: validator failed job logs

slickroot committed:
feat(subnet): enable all nvidia driver capabilities on docker containers

pyon12 committed:
Merge pull request #838 from Datura-ai/feature/enable-all-nvidia-capabilities-on-our-containers

feat(subnet): enable all nvidia driver capabilities on docker containers

pyon12 committed:
feat(validator): support NVIDIA H200 NVL and NVIDIA RTX PRO 6000

pyon12 committed:
chore(validator): bump version

arhangel66 committed:
feat(validator): DAH-1641 - add reconnect on case if connection to backend was failed.

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #840 from Datura-ai/feature/DAH-1641-fix-ws-reconnection

feat(validator): DAH-1641 - fix validator WebSocket reconnection on backend rejection

pyon12 committed:
Merge branch 'main' into feat/DAH-1632-remove-price-per-hour

pyon12 committed:
chore: bump versions

pyon12 committed:
Merge pull request #830 from Datura-ai/feat/DAH-1632-remove-price-per-hour

Feat: DAH-1632 - remove price per hour

pyon12 committed:
fix: conflcit and merge master

pyon12 committed:
feat(miner): skip validator verification for initial setup

pyon12 committed:
feat: add migration

pyon12 committed:
feat(miner): update validator hotkey in the config

pyon12 committed:
chore(miner): bump version

pyon12 committed:
Merge pull request #844 from Datura-ai/feat/DAH-1628-switch-valdaitor-hotkey

feat(miner): DAH-1628 - switch validator hotkey

pyon12 committed:
fix(validator): connector exceptin issue

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #846 from Datura-ai/feat/DAH-1628-switch-valdaitor-hotkey

fix(validator): connector exceptin issue

LegendWorrier committed:
feat: DAH-1623 - check executor ssh conenction before creating

fortunelucky777 committed:
enhance(validator/db): increase connection pool size and improve session logging

fortunelucky777 committed:
Merge branch 'main' into hotfix/db-connection-issue

fortunelucky777 committed:
refactor(validator): streamline connection handling and error logging in ComputeClient

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #848 from Datura-ai/hotfix/db-connection-issue

enhance(validator/db): increase connection pool size and improve session logging

fortunelucky777 committed:
fix(validator): adjust reconnect delay strategy in ComputeClient for improved stability

fortunelucky777 committed:
Merge branch 'hotfix/db-connection-issue' of github.com:Datura-ai/lium-io into hotfix/db-connection-issue

fortunelucky777 committed:
chore(validator): bump version to 3.13.22

pyon12 committed:
Merge pull request #851 from Datura-ai/hotfix/ws-retry

hotfix: ws retry

LegendWorrier committed:
Merge branch 'main' into feat/DAH-1623-check-executor-miner-side-before-create

pyon12 committed:
fix: git conflict

pyon12 committed:
feat(validator): update gpu model rate for H200 NVL and RTX PRO 6000

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #800 from Datura-ai/feat/DAH-1589-gpu-splitting

Feat: DAH-1589 gpu splitting

pyon12 committed:
feat(miner,validator): update required deposit amount

pyon12 committed:
fix: git conflict from master

pyon12 committed:
chore(miner,validator): bump versions

pyon12 committed:
Merge branch 'main' into feat/gpu-model-h200-nvl

pyon12 committed:
Merge pull request #839 from Datura-ai/feat/gpu-model-h200-nvl

Feat: support more gpu models

pyon12 committed:
hotfix: disable connecitivity check

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #855 from Datura-ai/hotfix/disable-connectivity-ceck-termperary

hotfix: disable connecitivity check

pyon12 committed:
feat(validator): add port connectivity check back

pyon12 committed:
fix(validator): cleaning process on the port connectivity check

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #857 from Datura-ai/hotfix/disable-connectivity-ceck-termperary

Hotfix/disable connectivity ceck termperary

pyon12 committed:
fix(validator): DAH-1648 - hardware utilization

pyon12 committed:
feat: scrape storage limit availability

pyon12 committed:
feat: DAH-1637 - implement storage/volume size limit

pyon12 committed:
fix(validator): score portion calculate if it's already set as 0

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #862 from Datura-ai/fix/score-portion-calc

fix(validator): score portion calculate if it's already set as 0

pyon12 committed:
chore(validator): rename to storage_limit_supported

pyon12 committed:
feat(validator): DAH-1637 - send back volume_limit_gb after container created for better tracking

MahmoudMohajer committed:
fix: Prevent SSH key substitution vulnerability in `upload_ssh_key` by validating public key against data to sign and add a new test for this fix.

MahmoudMohajer committed:
feat: import HTTPException from fastapi.

pyon12 committed:
feat(miner): cli and readme update

pyon12 committed:
Merge pull request #850 from Datura-ai/feat/DAH-1623-check-executor-miner-side-before-create

feat: DAH-1623 - check executor ssh conenction before creating

pyon12 committed:
chore(miner): bump version

pyon12 committed:
Merge pull request #865 from Datura-ai/feat/DAH-1623-check-executor-miner-side-before-create

chore(miner): bump version

pyon12 committed:
Merge pull request #864 from MahmoudMohajer/fix/744-public-key-validation

Fix SSH Key Substitution Vulnerability in Public Key Injection

pyon12 committed:
Merge pull request #860 from Datura-ai/fix/DAH-1648-hardware-utilization

fix(validator): DAH-1648 - hardware utilization

pyon12 committed:
chore(executor): bump version

pyon12 committed:
chore(miner): bump version

pyon12 committed:
feat: memory restriction field type to int from str

pyon12 committed:
chore(validator): bump version

pyon12 committed:
fix: git conflict and merge

pyon12 committed:
Merge pull request #861 from Datura-ai/feat/DAH-1637-docker-volume-restriction

Feat: DAH-1637 - docker  storage/volume restriction

pyon12 committed:
fix: renting-in-progress status when renting a pod

pyon12 committed:
feat(validator): limit the gpu price in synthetic job

pyon12 committed:
fix: scoring and message for price limit

pyon12 committed:
fix: scoring and message for price limit

pyon12 committed:
Merge pull request #873 from Datura-ai/hotfix/gpu-splitting-renting-in-progress

fix: renting-in-progress status when renting a pod

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #874 from Datura-ai/feat/gpu-price-limit

feat(validator): limit the gpu price in synthetic job

pyon12 committed:
fix(validator): issue when price_per_gpu is null

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #876 from Datura-ai/feat/gpu-price-limit

Feat/gpu price limit

pyon12 committed:
feat(executor): DAH-1655 - add storage/volume metrics

pyon12 committed:
feat(validator): add more logs for port range settings when port verification failed

pyon12 committed:
Merge pull request #879 from Datura-ai/feat/port-range-logs-when-failed

feat(validator): add more logs for port range settings

pyon12 committed:
feat(validator): add B300 GPU

pyon12 committed:
feat(validator): B300 model rate to 0.05

pyon12 committed:
fix: correct B300 gpu name

pyon12 committed:
Merge pull request #878 from Datura-ai/feat/DAH-1655-pod-storage-volume-metrics

feat(executor): DAH-1655 - add storage/volume metrics

pyon12 committed:
feat(validator): increase burn rate to 0.95

pyon12 committed:
Merge pull request #881 from Datura-ai/feat/burn-rate-95

feat(validator): increase burn rate to 0.95

pyon12 committed:
Merge pull request #880 from Datura-ai/feat/add-b300

feat(validator): add B300 GPU

pyon12 committed:
chore(validator,executor): bump versions

pyon12 committed:
chore(validator,executor): bump versions

pyon12 committed:
feat(validator): burn rate to 0.91

pyon12 committed:
chore(validator): bump version

pyon12 committed:
Merge pull request #884 from Datura-ai/feat/burn-rate-91

feat(validator): burn rate to 0.91

MkDev11 committed:
fix: use apt-get instead of apt in install scripts

apt is designed for interactive terminal use while apt-get is intended
for scripted/automated usage. apt's output format and behavior can change
between versions without warning, whereas apt-get has a stable CLI interface.

Fixes #198

pyon12 committed:
Merge pull request #886 from MkDev11/fix/use-apt-get-in-scripts

fix: use apt-get instead of apt in install scripts

MkDev11 committed:
fix: improve SSH key substitution vulnerability fix (Issue #744)

Improvements over PR #864:

1. Fix /remove_ssh_key endpoint - had the same vulnerability but was not fixed
2. Add whitespace normalization - keys with trailing newlines/spaces now match
3. Add security logging - log warnings when potential attacks are detected
4. Extract to reusable helper function - _validate_ssh_key_consistency()
5. Improve tests - add whitespace normalization and edge case tests

Security: Without this fix, an attacker could intercept a request and
substitute a different public_key while keeping the valid signature for
data_to_sign, gaining unauthorized SSH access.

Fixes #744

LegendWorrier committed:
feat: add new driver version 590.44.01

LegendWorrier committed:
chore: update the version

pyon12 committed:
Merge pull request #889 from Datura-ai/feat/add-new-driver-version-590.44.01

feat: add new driver version 590.44.01

pyon12 committed:
Merge pull request #888 from MkDev11/fix/ssh-key-substitution-vulnerability-744

fix: improve SSH key substitution vulnerability fix (Issue #744)

slickroot committed:
feat(validator): Allow a dry run mode to be able to run a second validator against real executors

slickroot committed:
fix(validator): close ssh connections on interactive shell service

slickroot committed:
fix(validator): we can uplaod the files, no problem

arhangel66 committed:
DAH-1676, add staging cd.

arhangel66 committed:
DAH-1676 simplify staging CD workflows

arhangel66 committed:
DAH-1676, update naming

arhangel66 committed:
DAH-1689, using backend API to get rented ports

arhangel66 committed:
DAH-1676, update tests

arhangel66 committed:
DAH-1689, fix validator-backend connection

arhangel66 committed:
DAH-1689, refactor validator services to use backend API for fetching rented executors

arhangel66 committed:
DAH-1689, fix tests

arhangel66 committed:
DAH-1689, rename "name" field to "container_name" in RentedPod and update references across the codebase.

arhangel66 committed:
DAH-1689, revert format changes

arhangel66 committed:
DAH-1689, revert format changes

arhangel66 committed:
DAH-1689, adapt banned gpu check for new way.

arhangel66 committed:
DAH-1689, add detailed HTTP GET/POST logging and reorganize rented executors fetching in validator logic.

slickroot committed:
fix(tests): fixed failed tests after checks were changed

slickroot committed:
Merge pull request #897 from Datura-ai/fix/failing-pipeline-tests

fix(tests): fixed failed tests after checks were changed

slickroot committed:
fix(tests): moved src/tests test cases to tests/, preparing for a github workflow

slickroot committed:
Merge pull request #898 from Datura-ai/fix/move-tests-to-single-directory

fix(tests): moved src/tests test cases to tests/, preparing for a github workflow

slickroot committed:
Merge pull request #895 from Datura-ai/feature/DAH-1676-prepare-staging-cd

DAH-1676, add staging cd.

slickroot committed:
feat(ci): added a tests workflow

slickroot committed:
Merge pull request #899 from Datura-ai/feat/github-workflow-tests

feat(ci): added a tests workflow

slickroot committed:
feat(subnet): lighter port mapping

slickroot committed:
Merge pull request #894 from Datura-ai/feat/validator-dry-mode

Feat/validator dry mode

slickroot committed:
chore(subnet): clean dead code

slickroot committed:
Merge pull request #900 from Datura-ai/feat/DAH-1690-better-port-mapping-check

feat(subnet): lighter port mapping

slickroot committed:
feat(subnet): added a new service validator-dry-run to run one loop with no fastAPI to test validation logic on prod

slickroot committed:
fix(subnet): unique name for dry run validator

arhangel66 committed:
DAH-1689, remove legacy test cases and adjust remaining tests for port mapping refactor.

arhangel66 committed:
DAH-1689, return original import string.

slickroot committed:
fix(subnet): no port forwarding needed

slickroot committed:
Merge pull request #901 from Datura-ai/feat/dry-run-in-prod

feat(subnet): added a new service validator-dry-run to run one loop w…

slickroot committed:
fix(subnet): port connectivity making one container per batch, and having access to logs

Commit:
security: Fix command injection vulnerabilities

- Apply shlex.quote() to all user inputs in shell commands
- Remove shell=True from subprocess calls
- Add input sanitization for startup_commands
- Fix vulnerabilities in SSH, Docker, and subprocess operations

See SECURITY_IMPROVEMENTS.md for detailed information.

arhangel66 reviewed (changes requested):
Thanks for tackling this! The direction is right, but a few issues need fixing before merge.

### Critical: double-quoting breaks commands

When using `subprocess.run(..., shell=False)` with a list, `shlex.quote()` should not be used — the list form is already safe. Adding quotes inserts literal quote characters into values:

```python
# Current (broken)
source_volume_quoted = shlex.quote(args.source_volume)  # becomes '/data/vol'
command = ["docker", "run", "-v", f"{source_volume_quoted}:{path_quoted}"]
# Docker receives: -v '/data/vol':'/mnt' — fails

# Correct
command = ["docker", "run", "-v", f"{args.source_volume}:{args.source_volume_path}"]
```

Affected: `backup_storage.py:124-145`, `restore_storage.py:52-68`

Same issue with nested quoting in `docker_service.py:481-498, 521-525` — values are quoted twice.

### Important: imports inside functions

`import shlex` is added inside functions in several files (`machine_scrape.py:609`, `verifyx_validation_service.py:148`, `rented_machine.py:209`, `executor_connectivity_service.py:539`). Should be at file top per project conventions.

### Suggestion: centralized helper

Instead of scattered `shlex.quote()` calls, consider a single tested helper:

```python
# src/services/command_utils.py
import shlex

def prepare_shell_command(template: str, **kwargs) -> str:
    """For SSH commands or shell=True. List-based subprocess needs no escaping."""
    escaped = {k: shlex.quote(str(v)) for k, v in kwargs.items()}
    return template.format(**escaped)
```

Usage:
```python
command = prepare_shell_command(
    "/usr/bin/docker exec {container} sh -c 'apt-get update'",
    container=container_name
)
```

This eliminates double-quoting mistakes and keeps security logic in one testable place + add a minimal tests for it.

slickroot committed:
Merge pull request #902 from Datura-ai/fix/connectivity-service-fixes

fix(subnet): port connectivity making one container per batch, and ha…

Commit:
security: Fix double-quoting and move imports to top

- Remove shlex.quote() from list-based subprocess calls (backup_storage.py, restore_storage.py)
- Fix nested quoting issues in docker_service.py
- Move import shlex from function bodies to file top (per project conventions)
- Add centralized command_utils.py helper with tests

Fixes review feedback on command injection security improvements.

Commit:
security: Address review feedback on command injection fixes

- Fix double-quoting in subprocess list calls
- Move imports to file top per conventions
- Add command_utils helper with tests

sm4rtdev committed:
Merge branch 'Datura-ai:main' into security/fix-command-injection-vulnerabilities

slickroot committed:
fix: switched to fallback, small refactoring to the service

slickroot committed:
fix(connectivity): a rewrite to the connectivity service, much simpler contracts and responsibilities, with very readable unit tests

slickroot committed:
fix: have the check log results any way

slickroot committed:
fix: remove protocol usage and used pytest.mocker on service tests

slickroot committed:
fix: renamed dind.py to dind_probe.py

slickroot committed:
fix: renamed dind.py to dind_probe.py

slickroot committed:
fix: renamed dind.py to dind_probe.py

sm4rtdev committed:
Merge branch 'main' into security/fix-command-injection-vulnerabilities
